### PR TITLE
Core: Avoid extra getFileStatus call in HadoopInputFile

### PR DESCRIPTION
**Pull request by singhpk234**
### About the change

In case we call both exists() & getLength() we make a redundant call to `getFileStatus` as `fs.exists(path)` internally checks 
```java
  /** Check if exists.
   * @param f source file
   */
  public boolean exists(Path f) throws IOException {
    try {
      return getFileStatus(f) != null;
    } catch (FileNotFoundException e) {
      return false;
    }
  }
```

So we could getFileStatus directly and cache it via lazyStat() in `stat` prop and in `getLength` & subsequent calls directly return it.

Note : This similar opt exists in S3InputFile as well as we cache the results in `getObjectMetadata` call https://github.com/apache/iceberg/blob/b9bbcfb0e4d8f2605d3c6fb2543cefdd5d09524d/aws/src/main/java/org/apache/iceberg/aws/s3/BaseS3File.java#L70

----

### Testing Done 

Existing UT's

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/hadoop/HadoopInputFile.java [hunk 1]
```diff
@@ -224,9 +224,13 @@ public String location() {
   @Override
   public boolean exists() {
     try {
-      return fs.exists(path);
-    } catch (IOException e) {
-      throw new RuntimeIOException(e, "Failed to check existence for file: %s", path);
+      return lazyStat() != null;
+    } catch (RuntimeIOException re) {
+      if (re.getCause() instanceof FileNotFoundException) {
```
Comment (szehon-ho) [inline core/src/main/java/org/apache/iceberg/hadoop/HadoopInputFile.java:229]: Question, what do you think, about making lazyStat() catch FileNotFoundException specifically and throwing Iceberg's NotFoundException in that case?  Then here we can catch directly NotFoundException?

**Review by szehon-ho — COMMENTED** [review]
I think it's a good idea, just a thought here

File: core/src/main/java/org/apache/iceberg/hadoop/HadoopInputFile.java (hunk 1 above)
Comment (singhpk234) [reply core/src/main/java/org/apache/iceberg/hadoop/HadoopInputFile.java]: I think it makes sense :) , made the corresponding changes !

File: core/src/main/java/org/apache/iceberg/hadoop/HadoopInputFile.java [hunk 2]
```diff
@@ -160,6 +160,8 @@ private FileStatus lazyStat() {
     if (stat == null) {
       try {
         this.stat = fs.getFileStatus(path);
+      } catch (FileNotFoundException e) {
+        throw new NotFoundException(e, "Path does not exist: %s", path);
```
Comment (szehon-ho) [inline core/src/main/java/org/apache/iceberg/hadoop/HadoopInputFile.java:164]: Nit: maybe more correct: "File does not exist"?

**Review by szehon-ho — COMMENTED** [review]
Sorry it may be silly, but what do you think to add a quick test in HadoopFileIOTest?  To catch regression if Hadoop ever change exception type.  I guess should be quick if it does not exist.  Or let me know if you think there's already coverage.

**Comment by singhpk234** [comment]
> but what do you think to add a quick test in HadoopFileIOTest?

exists() coverage is covered in `TestStructuredStreamingRead3` and `TestDropTable`, but it defintely makes sense to add one in HadoopFileIO as well. made the changes.

**Review by szehon-ho — APPROVED** [review]
Looks good to me, thanks for the test!

**Review by jackye1995 — APPROVED** [review]
Looks good to me as well

**Comment by szehon-ho** [comment]
Thanks @singhpk234 for change, and @jackye1995 for extra review

**Pull request merged**